fairlearn/metrics/_metric_frame.py

```diff
@@ -194,24 +207,181 @@ def __init__(self,
+
         y_t = _convert_to_ndarray_and_squeeze(y_true)
         y_p = _convert_to_ndarray_and_squeeze(y_pred)
 
-        func_dict = self._process_functions(metrics, sample_params)
+        self._y_true.append(y_t)
+        self._y_pred.append(y_p)
+        self._sensitive_features.append(sensitive_features)
+        if self._control_features is not None:
+            check_consistent_length(y_true, control_features)
+            self._control_features.append(control_features)
+        elif control_features is not None:
+            raise ValueError(_CF_BAD_STATE_ERR)
+
+        if self._sample_params is not None:
+            if sample_params is not None:
+                # All arrays should be of equal length
+                check_consistent_length(y_true, *sample_params.values())
+                self._sample_params = self._concat_batches([self._sample_params, sample_params])
+            else:
+                raise ValueError("MetricFrame expected `sample_params` to be supplied.")
+
+        # Reset metric
+        self._overall = None
+        self._by_group = None
+
+    def _concat_batches(self, batches):
+        """Concatenate a list of items together.
+
+        When merging Dicts, the values will be merged and a new dict will be created.
+        DataFrames' columns need to match to be concatenated.
+        Arrays, list and Series can be merged using simple functions.
 
+        Parameters
+        ----------
+        batches : List[Union[Dict, nd.array, List, Serie, DataFrame]]
+            A list of batch to concat together. The batches must be of the same type.
+
+        """
+        if len(batches) == 0:
+            raise ValueError(_EMPTY_BATCHES_ERR)
+        batch_type = type(batches[0])
+        if not all([type(arr) is batch_type for arr in batches]):
+            raise ValueError("Can't concatenate arrays of different types.")
+        # Simple cases, array, Series and list can be concat easily.
+        if batch_type is np.ndarray:
+            result = np.concatenate(batches)
+        elif batch_type is list:
+            result = sum(batches, [])
+        elif batch_type is pd.Series:
+            result = pd.concat(batches)
+        elif batch_type is pd.DataFrame:
+            # To concat dataframe, we check if the columns match.
+            col_nums = batches[0].columns
+            if any([col_nums != df.columns for df in batches]):
+                raise ValueError(f"Column mismatch expected {col_nums},"
+                                 f" got {[df.columns for df in batches]}")
+            result = pd.concat(batches)
+        elif batch_type is dict:
+            # Create a new dict that merges values of all keys.
+            all_keys = set(sum((list(b.keys()) for b in batches), []))
+            return {k: self._concat_batches([b[k] for b in batches if k in b])
+                    for k in all_keys}
+        else:
+            raise ValueError(f"Can't concatenate {batches}")
```

Comment:
I think I'd be happier if we just convert everything to numpy arrays and concatenate them instead of all these different cases. Do we need them to stay the same type?